### PR TITLE
[tools/circle_plus_gen] Add training parameters json loader 

### PR DESCRIPTION
**Pull request by zetwhite**
This PR add a json loader to parse a json file of training hyper parameters and construct TrainParam instance. 

ONE-DCO-1.0-Signed-off-by: seunghui youn <sseung.youn@samsung.com> 

* related issue : https://github.com/Samsung/ONE/issues/12650

### CODE REVIEW

File: tools/circle_plus_gen/main.py [hunk 1]
```diff
@@ -32,9 +38,32 @@ def print_training_hparameters(in_circle_file) -> typing.NoReturn:
...
+    '''
+    Inject hparams_file's contents into in_file's circle model, and save it as out_file 
+    '''
+    # if out_file isn't given, rewrite in_file
+    if out_file is None:
+        out_file = in_file
+
+    tparams: TrainParam = TrainParam.from_json(hparams_file)
+    print("load traing hyperparameters")
+    print(tparams.dump_as_json())
```
Comment (zetwhite) [inline tools/circle_plus_gen/main.py:51]: we could check that it sucessfully load the json files. 

```bash
$ python3 main.py example/sample_tpram.circle example/train_tparam.json  
load traing hyperparameters
{
    "optimizer": {
        "type": "adam",
        "args": {
            "learningRate": 0.01,
            "beta1": 0.9,
            "beta2": 0.999,
            "epsilon": 1e-07
        }
    },
    "loss": {
        "type": "categorical crossentropy",
        "args": {
            "fromLogits": true,
            "reduction": "sum over batch size"
        }
    },
    "batchSize": 32
}
````

**Comment by zetwhite** [comment]
I'll open this PR after https://github.com/Samsung/ONE/pull/13011 merged

**Comment by zetwhite** [comment]
@jyoungyun I explicitly asked you to review.  Since you've done similar work in this draft - https://github.com/Samsung/ONE/pull/12984, I guess you could give me a practical review.

File: tools/circle_plus_gen/main.py [hunk 2]
```diff
@@ -32,9 +38,32 @@ def print_training_hparameters(in_circle_file) -> typing.NoReturn:
...
+def inject_hparams(in_file, hparams_file, out_file=None) -> None:
+    '''
+    Inject hparams_file's contents into in_file's circle model, and save it as out_file 
+    '''
+    # if out_file isn't given, rewrite in_file
+    if out_file is None:
+        out_file = in_file
+
+    tparams: TrainParam = TrainParam.from_json(hparams_file)
+    print("load traing hyperparameters")
```
Comment (ragmani) [inline tools/circle_plus_gen/main.py:50]: Maybe a typo
```suggestion
    print("load training hyperparameters")
```

File: tools/circle_plus_gen/example/train_tparam.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      "type": "adam",
+      "args": {
+          "learningRate": 0.01,
+          "beta1": 0.9,
+          "beta2": 0.999,
+          "epsilon": 1e-07
+      }
+  },
+  "loss": {
+      "type": "categorical crossentropy",
```
Comment (jyoungyun) [inline tools/circle_plus_gen/example/train_tparam.json:12]: (question)
Where can I get the type string?

**Comment by jyoungyun** [comment]
This PR looks good to me.

But there is one concern. This tool requires a json file. And a json file contains multiple strings and arguments. I couldn't grasp json file format for this tool at a glance. I use input paramters to get hyper parameters from the draft code. The use of this draft is not very clear either. If we introduce this PR to use the json file as input to this tool, it would be better to provide a guideline for the json file format.

**Comment by zetwhite** [comment]
> it would be better to provide a guideline for the json file format.

I agree. 
I'd better add a guideline after the basic feature of this tool added. 

>  I use input paramters to get hyper parameters from the draft code. 

I also think it is good to extend this tool to get parameters through the input options.

File: tools/circle_plus_gen/example/train_tparam.json (hunk 3 above)
Comment (zetwhite) [reply tools/circle_plus_gen/example/train_tparam.json:12]: There is no document for type string rules. I'll add a document and more example files for details. 

-------------------------------

For now, you could check the allowable type string only in code level. 

https://github.com/Samsung/ONE/blob/9fe820c3e05eedec50edf4039a66dfb26d2fb322/tools/circle_plus_gen/lib/utils.py#L38

File: tools/circle_plus_gen/main.py [hunk 4]
```diff
@@ -32,9 +38,32 @@ def print_training_hparameters(in_circle_file) -> typing.NoReturn:
...
+    '''
+    Inject hparams_file's contents into in_file's circle model, and save it as out_file 
+    '''
+    # if out_file isn't given, rewrite in_file
+    if out_file is None:
+        out_file = in_file
+
+    tparams: TrainParam = TrainParam.from_json(hparams_file)
+    print("load training hyperparameters")
+    print(tparams.dump_as_json())
```
Comment (zetwhite) [inline tools/circle_plus_gen/main.py:51]: .

Comment (zetwhite) [reply tools/circle_plus_gen/main.py:51]: We could check that it sucessfully load the json files. 

```bash
$ cd tools/circle_plus_gen
$ python3 main.py example/sample_tpram.circle example/train_tparam.json  
load traing hyperparameters
{
    "optimizer": {
        "type": "adam",
        "args": {
            "learningRate": 0.01,
            "beta1": 0.9,
            "beta2": 0.999,
            "epsilon": 1e-07
        }
    },
    "loss": {
        "type": "categorical crossentropy",
        "args": {
            "fromLogits": true,
            "reduction": "sum over batch size"
        }
    },
    "batchSize": 32
}
````

**Review by jyoungyun — APPROVED** [review]
LGTM

**Review by ragmani — APPROVED** [review]
LGTM

**Pull request merged**